optuna/integration/wandb.py

```diff
@@ -0,0 +1,127 @@
+
+        Add Weights & Biases callback to Optuna optimization.
+
+        .. code::
+
+            import optuna
+            from optuna.integration.wandb import WeightsAndBiasesCallback
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -10, 10)
+                return (x - 2) ** 2
+
+
+            wandbc = WeightsAndBiasesCallback(
+                project_name="my-project-name",
+                group_name="my-group-name",
+                job_type="my-job-type",
+            )
+
+
+            study = optuna.create_study(study_name="my_study")
+            study.optimize(objective, n_trials=10, callbacks=[wandbc])
+
+    Args:
+        metric_name:
+            Name of the optimized metric. Since the metric itself is just a number,
+            ``metric_name`` can be used to give it a name. So you know later
+            if it was roc-auc or accuracy.
+        project_name:
+            Name of the project under which run should be categorized.
+            If the project is not specified, the run is put in an "Uncategorized" project.
+        group_name:
+            Specifies a group under project into which run should be categorized.
+            Please refer to `Group Runs
+            <https://docs.wandb.ai/guides/track/advanced/grouping>`_ for more details.
+        entity_name:
+            Username or team name under which run should be logged. When set to :obj:`None`,
+            the run is logged under the current user.
+        run_name:
+            Specifies the display name in Weights & Biases UI for this run.
+            When set to :obj:`None`, random two word name is generated instead.
+        job_type:
+            Specifies type of the run, and is used as additional grouping
+            for the runs.
+        mode:
+            Specifies execution mode for Weights & Biases. One of
+            ``online``, ``offline`` or ``disabled``. Defaults to online.
+    """
+
+    def __init__(
+        self,
+        metric_name: Union[str] = "value",
+        project_name: Optional[str] = None,
+        group_name: Optional[str] = None,
+        entity_name: Optional[str] = None,
+        run_name: Optional[str] = None,
+        job_type: Optional[str] = None,
+        mode: Union[str] = "online",
+    ) -> None:
+
+        _imports.check()
+
+        self._metric_name = metric_name
+        self._project_name = project_name
+        self._group_name = group_name
+        self._entity_name = entity_name
+        self._run_name = run_name
+        self._job_type = job_type
+        self._mode = mode
+
+        self._initialize_run()
+
+    def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> None:
+
+        trial_state = str(trial.state).split(".")[-1]
+        direction = str(study.direction).split(".")[-1]
+
+        attributes = {
+            "direction": direction,
+            "trial_state": trial_state,
+            "distributions": trial.distributions,
```

Comment:
`trial.distributions` can be different among trials (I give an example below), so could you store `trial.distributions` by using `wandb.log`?

```python
import optuna
from optuna.integration.wandb import WeightsAndBiasesCallback


def objective(trial):
    if trial.number % 2:
        x = trial.suggest_float("x", -10, 10)
    else:
        x = trial.suggest_float("x", -10, 20)
    y = trial.suggest_float("y", -10, 10)
    return (x - 2) ** 2 + y


wandbc = WeightsAndBiasesCallback()
study = optuna.create_study(study_name="my_study")
study.optimize(objective, callbacks=[wandbc], n_trials=10)

```

Reply:
I think `wandb.log` is meant to be used with key-value pairs, where values are either scalars or one of [wandb artifacts](https://docs.wandb.ai/ref/python/data-types). If we pass dict with strings, it will not be accessible, instead rendering following information:

![image](https://user-images.githubusercontent.com/37713008/124550676-4fe36b00-de31-11eb-875c-2f78dbd9c34a.png)

Proper way to store this, and other trial attributes seems to be an artifact [table](https://docs.wandb.ai/ref/python/data-types/table), where primary key is the trial number. But there is another problem - `wandb.log` does't really seem to handle well continuous logging of updated artifact. Consider following example:

If we roll out couple calls to our callback, including constant updates of our artifact, and giving more than enough time between `log` calls, we will end up looking with something like this

```python
import wandb
import time

wandb.init()

table = wandb.Table(columns=['foo', 'bar'])

table.add_data(*[0, 0])
time.sleep(1.)
wandb.log({"metric": 1, "my-table": table}, step=1)

table.add_data(*[1, 1])
time.sleep(1.)
wandb.log({"metric": 2, "my-table": table}, step=2)

table.add_data(*[2, 2])
time.sleep(1.)
wandb.log({"metric": 3, "my-table": table}, step=3)

table.add_data(*[3, 3])
time.sleep(1.)
wandb.log({"metric": 4, "my-table": table}, step=4)
```

which looks like that UI side

![image](https://user-images.githubusercontent.com/37713008/124551828-0267fd80-de33-11eb-96dd-38bdf7362b53.png)

Note that we are missing some rows - I think `wandb` just does not manage to update artifact on time. This is even mentioned in `wandb.log` documentation:
> wandb.log is not intended to be called more than a few times per second. If you want to log more frequently than that it's better to aggregate the data on the client side or you may get degraded performance.

This can be fixed easily, by just building a table throughout the run, and commiting it at the end like so

```python
# [...]

table.add_data(*[0, 0])
time.sleep(1.)
wandb.log({"metric": 1}, step=1)

table.add_data(*[1, 1])
time.sleep(1.)
wandb.log({"metric": 2}, step=2)

table.add_data(*[2, 2])
time.sleep(1.)
wandb.log({"metric": 3}, step=3)

table.add_data(*[3, 3])
time.sleep(1.)
wandb.log({"metric": 4}, step=4)

wandb.log({"my-table": table})
``` 

but unfortunately, we don't really have a way to tell our callback that optimization is finished and it should upload artifacts (hence my earlier comments about allowing callbacks to use initialization and cleanup methods like e.g. [keras callback](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback) `on_train_{begin | end}`).

So, due to all this, i think as a temporary solution, distributions should stay logged as part od config, so they can render properly in main run table, and to address your example - i can make callback track distributions used throughout the study, and just update config when anything new shows up. Please let me know if this is ok with you.

Also, for `trial.user_attr` same problems as above apply, but due to the fact that it can be any `dict`, even nested I'm not sure if  `config` workaround will work in this case, since attributes would get unrolled into main table, thus cluttering it. Can we not make it part of this PR, and wait for proper handling of artifacts? Sorry for bringing up `tral.user_attr` here, not in the original comment, but story is the same for all trial attributes.

Reply:
Awesome! Thank you for sharing your investigation. Alternatively, we can ignore `distributions` at this time–do not track them by using wandb. `tral.user_attr` as well for the same reason as you provided.

Reply:
Great, I will remove all trial related attributes for the time being (so `distributions`, `user_attr` and `state`). As I said previously I can propose and work on extending callback API if you guys don't mind such changes, and come back to improve this callback with proper implementation of artifact tracking.

Reply:
Sounds perfect!

Reply:
Addressed in 5b17d4e0de5b085dbea8cc2d703857b9e835e76e